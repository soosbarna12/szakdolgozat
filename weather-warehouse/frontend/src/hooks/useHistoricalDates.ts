import { useQuery } from "@tanstack/react-query";
import dayjs from "dayjs";
import { useEffect } from "react";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";

export function useHistoricalDates(location: string) {
  const { showAlert } = useAlert();

  const { data, error, isLoading } = useQuery({
    queryKey: ["historicalDates", location],
    queryFn: async () => {
      const response = await axios.get(`/api/historical/historicalDates`, {
        params: { location },
      });
      
      return response.data.map((date: any) => dayjs(date.date).format("YYYY-MM-DD"));
    },
    
    enabled: !!(location),
    refetchOnWindowFocus: false,
  });

  useEffect(() => {
    if (error) {
      showAlert("Error fetching historical dates", "error");
    }
  }, [error]);

  return { data: data || [], error, isLoading };
}
