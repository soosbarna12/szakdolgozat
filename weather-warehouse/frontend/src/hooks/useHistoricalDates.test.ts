import '@testing-library/jest-dom';
import { waitFor } from '@testing-library/react';
import axios from '../utils/axiosConfig';
import { renderHookWithQueryClient } from '../utils/test/renderHookWithQueryClient';
import { useHistoricalDates } from './useHistoricalDates';

jest.mock('../utils/axiosConfig');

describe('useHistoricalDates', () => {
  const mockResponseData = [{date: '2023-01-01'}, {date: '2023-01-02'}, {date: '2023-01-03'}];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches useHistoricalDates hook snapshot', async () => {
    (axios.get as jest.Mock).mockResolvedValueOnce({ data: mockResponseData });

    const { result } = renderHookWithQueryClient(() => useHistoricalDates('New York'));

    await waitFor(() => {
      expect(result.current.data).toEqual(mockResponseData.map((date: any) => date.date));
    });

    expect(result.current).toMatchSnapshot();
  });

  it('fetches historical dates successfully', async () => {
    (axios.get as jest.Mock).mockResolvedValueOnce({ data: mockResponseData });

    const { result } = renderHookWithQueryClient(() => useHistoricalDates('New York'));

    await waitFor(() => {
      expect(result.current.data).toEqual(mockResponseData.map((date: any) => date.date));
      expect(result.current.isLoading).toBe(false);
      expect(result.current.error).toBeNull();
    });
  });

  it('returns an empty array when no data is found', async () => {
    (axios.get as jest.Mock).mockResolvedValueOnce({ data: [] });

    const { result } = renderHookWithQueryClient(() => useHistoricalDates('New York'));

    await waitFor(() => {
      expect(result.current.data).toEqual([]);
      expect(result.current.isLoading).toBe(false);
      expect(result.current.error).toBeNull();
    });
  });

  it('does not fetch data when location is empty', async () => {
    const { result } = renderHookWithQueryClient(() => useHistoricalDates(''));

    await waitFor(() => {
      expect(result.current.data).toEqual([]);
      expect(result.current.isLoading).toBe(false);
      expect(result.current.error).toBeNull();
    });

    expect(axios.get).not.toHaveBeenCalled();
  });
});