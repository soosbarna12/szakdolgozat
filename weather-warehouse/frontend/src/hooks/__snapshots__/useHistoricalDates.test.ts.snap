// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useHistoricalDates matches useHistoricalDates hook snapshot 1`] = `
{
  "data": [
    "2023-01-01",
    "2023-01-02",
    "2023-01-03",
  ],
  "error": null,
  "isLoading": false,
}
`;
